thorlcr/thorutil/thbuf.cpp

```diff
@@ -2493,21 +2481,23 @@ class CSharedFullSpillingWriteAhead : public CInterfaceOf<ISharedRowStreamReader
     {
         if (outputStream)
         {
+            outputStream.clear();
             iFileIO->flush();
             tempFileOwner->noteSize(iFileIO->getStatistic(StSizeDiskWrite));
             ::mergeStats(inactiveStats, iFileIO);
             iFileIO.clear();
-            outputStream.clear();
         }
     }
     void createOutputStream()
     {
+        closeWriter(); // Ensure stats from closing files are preserved in inactiveStats
         // NB: Called once, when spilling starts.
         tempFileOwner.setown(activity.createOwnedTempFile(baseTmpFilename));
         auto res = createSerialOutputStream(&(tempFileOwner->queryIFile()), compressHandler, options, numOutputs + 1);
         outputStream.setown(std::get<0>(res));
         iFileIO.setown(std::get<1>(res));
         totalInputRowsRead = inMemTotalRows;
+        inactiveStats.addStatistic(StNumSpills, 1);
```

Comment:
hm, in previous commit it performed a : 
```
::mergeStats(inactiveStats, iFileIO);
```
.. here. But as Gavin pointed out:
>I think this is going to double count the stats. Each time it is called it will add the stats from the previous call as well as the rows just written.

But not with it removed, it will not update stats for the spill until the closeWriter, i.e. when at very end.
Really, tt should (as it was) update in writeRowsFromInput, so we see progress, this is particularly important in this spill context (less so in e.g. sort spilling), because these tend to be quite long lived (NB: we should revisit all so they are periodic though, not at end of file).

Perhaps use updateDelta here (you would also need to do so in closeWriter if so)

Reply:
I have updated getStatistic to return the statistic from iFileIO (if it is still open) and from inactiveStats (thbuff.cpp:2717) .  I believe the in-progress stats would be updated. @jakesmith

Reply:
See other new/old comment [ https://github.com/hpcc-systems/HPCC-Platform/pull/18819#discussion_r1673795535 ]

Examining iFileIO in getStatistic is thread unsafe, which was the reason to move away from that implementation.